### PR TITLE
Users should be able to signup

### PR DESCRIPTION
**Pull request by azaria-dedmon**
**What does this PR do?**

- This PR makes signup functionality available for users.

**Description of Task to be completed**

- Create "/register" endpoint
- Create "/user" endpoint to be able to test if the user is redirected to the appropriate page after registering.
- Add "signup" method to "User" table.
- Write relevant unit tests.

**How should this be manually tested?**

- Run `python3 run.py` in the terminal.

- Go to `localhost:5000/register`

### CODE REVIEW

File: app.py [hunk 1]
```diff
@@ -1,6 +1,49 @@
...
-from models import db, connect_db
+from flask import Flask, session, g, render_template, redirect, request
+from forms import RegisterUser
+from models import db, connect_db, User, testing_states
+
+
+CURR_USER_KEY = "curr_user"
 
 app = Flask(__name__)
+app.config['SECRET_KEY'] = "application20212021"
```
Comment (alexkayabula) [inline app.py]: @azaria-dedmon use environment variables to secure application secrets.

**Review by alexkayabula — CHANGES_REQUESTED** [review]

**Comment by alexkayabula** [comment]
@azaria-dedmon write tests for all untested lines of code. It is important we have all code tested as we strive for 100% test coverage.

**Comment by alexkayabula** [comment]
@azaria-dedmon please set a `secret key` in the tests setup to fix the `RuntimeError: Session backend did not open a session. Check the configuration` error.

**Comment by azaria-dedmon** [comment]
@alexkayabula What exactly are you referring to by settings?

**Comment by alexkayabula** [comment]
@azaria-dedmon sorry for the misleading comment. I have updated it. Kindly re-check it.

**Comment by alexkayabula** [comment]
@azaria-dedmon since we are using a mock secret in the tests, use an actual string instead of an environment variable.

**Comment by azaria-dedmon** [comment]
@alexkayabula that seemed to have solved the issue. Thanks

**Comment by azaria-dedmon** [comment]
@alexkayabula Hey I seem to be having issues with getting my code to pass again. Everything is passing with an **OK** locally, however, when I click on the coverage report from Coveralls, there is still 3 lines of code that seem to be failing. This particular line of code causes an IntegrityError and I am testing for that in my code. I cant seem to figure out why it's not passing on Coveralls.

**Comment by azaria-dedmon** [comment]
@alexkayabula Nevermind my previous comment; I figured out what the issue was.

**Comment by azaria-dedmon** [comment]
The `/user` endpoint as well as `user_homepage` template were included in this PR in order to be able to test if the user is redirected to the appropriate page after registering.

**Comment by alexkayabula** [comment]
@azaria-dedmon what do you mean by `Run app.py` in the description ? Specify the actual command to run the application.

**Comment by alexkayabula** [comment]
@azaria-dedmon Which database do you use for the application ? I can't find any database setup in the codebase.

**Comment by azaria-dedmon** [comment]
@alexkayabula Whenever I run my application, I always use `flask run` in the command line. Would the same apply for other's trying out the application?

**Comment by azaria-dedmon** [comment]
@alexkayabula Initially I had included a datbase named, 'covidtest' in app.py. However, when the Travis CI tests failed, you required that it be removed in order for the tests to stop throwing errors.

**Comment by alexkayabula** [comment]
> @alexkayabula Whenever I run my application, I always use `flask run` in the command line. Would the same apply for other's trying out the application?

Yes it does.

**Comment by alexkayabula** [comment]
> @alexkayabula Initially I had included a datbase named, 'covidtest' in app.py. However, when the Travis CI tests failed, you required that it be removed in order for the tests to stop throwing errors.

I do remember. Let me try to come up with a way of adding the config without failing the tests.

File: app.py [hunk 2]
```diff
@@ -1,6 +1,59 @@
...
+                image=form.image.data,
+                state=form.state.data,
+                vax_date=form.vax_date.data,
+                covid_status=form.covid_status.data)
+            db.session.commit()
+        except IntegrityError:
+            flash("Invalid information", 'danger')
+            return render_template('users/register.html', form=form)
+        do_login(user)
+        return redirect('/user')
```
Comment (alexkayabula) [inline app.py]: @azaria-dedmon Isn't this supposed to be in the `try` block ? What happens when a user registers successfully ?

Comment (azaria-dedmon) [reply app.py]: I am using the try block to test specifically on whether or not the user is able to actually register for an account. I still have the action of the user's redirect nested inside of the if statement. Would it be considered best practice to include the redirect inside of the try block as well?

**Comment by alexkayabula** [comment]
@azaria-dedmon please resolve merge conflicts and confirm everything works fine.

File: app.py (hunk 2 above)
Comment (alexkayabula) [reply app.py]: I guess if the user successfully registers, the `do_login` function is called then a redirect occurs. If that is the case, then the `do_login` function should be in the `try` block.

**Comment by alexkayabula** [comment]
@azaria-dedmon please squash the 2 commits into one commit and name it appropriately.

**Comment by azaria-dedmon** [comment]
@alexkayabula 
Would it be just as affective to delete both commits and just push another commit up? I am having difficulty squashing

**Comment by alexkayabula** [comment]
> @alexkayabula
> Would it be just as affective to delete both commits and just push another commit up? I am having difficulty squashing

There is no need to delete.

**Comment by azaria-dedmon** [comment]
@alexkayabula My tests are now failing and I keep getting an import error. I have been trying to fix this bug since early yesterday morning but cannot figure out why my unit test, "test_user_views.py" keeps breaking.

**Comment by alexkayabula** [comment]
> @alexkayabula My tests are now failing and I keep getting an import error. I have been trying to fix this bug since early yesterday morning but cannot figure out why my unit test, "test_user_views.py" keeps breaking.

Let me check it out.

**Comment by alexkayabula** [comment]
@azaria-dedmon review and test out  this pull request and confirm everything works fine.

**Comment by azaria-dedmon** [comment]
@alexkayabula it seems to be a local issue because all of the tests are passing here. I'm going to try and figure out what's going on with my computer

**Comment by alexkayabula** [comment]
> @alexkayabula it seems to be a local issue because all of the tests are passing here. I'm going to try and figure out what's going on with my computer

@azaria-dedmon add this issue under the [issues](https://github.com/azaria-dedmon/covid-19/issues) section. I will look into it.

**Comment by alexkayabula** [comment]
LGTM

**Pull request merged**